### ISSUE TITLE
How to count the number of multiplications given a keras model?

### ISSUE THREAD

**Issue by JustinYou1992** [fchollet/keras#4339]
Hi,

I am wondering how to count the number of multiplications given a keras model, which is well defined. Is there any off-the-shelf function or codes?

I know obtaining the number of parameters given a keras model is to use `model.summary()` .

**Comment by carlthome** [comment]
I don't know what particular model you're using but isn't the number of parameters pretty much the same as how many multiplications/divisions you'd expect (aside from some additions here and there but they're probably proportionally few even for a ResNet)?

**Comment by JustinYou1992** [comment]
@carlthome thx for comments. Let's talk about VGG16net. I want to know how many parameters in this net and I can use `model.summary()` to directly display the total number of this model's parameters. Now I wanna know given a input, how many multiplications would be involved in one-pass of the whole net....

Is there any function capable of doing this calculation?

**Comment by carlthome** [comment]
No, AFAIK there's no such function. For VGG16 it depends on how the convolutions are implemented in CUDA, which in turns depends on the Keras backend. You could assume the fully-connected layers have as many multiplications as weights (plus bias), and that the pooling layers have none. The convolution layers are probably done with FFT in which case there would be as many multiplications as parameters ([I guess!](https://en.wikipedia.org/wiki/Convolution_theorem)).

**Comment by JustinYou1992** [comment]
Yes. The real implementation depends  on the backend. Maybe I can calculate it approximately just like calculating the computational complexity of some network. 
            On 11/15/2016 23:56, Carl Thomé wrote:No, AFAIK there's no such function. For VGG16 it depends on how the convolutions are implemented in CUDA, which in turns depends on the Keras backend. You could assume the fully-connected layers have as many multiplications as weights (plus bias), and that the pooling layers have none. The convolution layers are probably done with FFT in which case there would be as many multiplications as parameters (I guess!).

—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/fchollet/keras","title":"fchollet/keras","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/fchollet/keras"}},"updates":{"snippets":[{"icon":"PERSON","message":"@carlthome in #4339: No, AFAIK there's no such function. For VGG16 it depends on how the convolutions are implemented in CUDA, which in turns depends on the Keras backend. You could assume the fully-connected layers have as many multiplications as weights (plus bias), and that the pooling layers have none. The convolution layers are probably done with FFT in which case there would be as many multiplications as parameters ([I guess!](https://en.wikipedia.org/wiki/Convolution_theorem))."}],"action":{"name":"View Issue","url":"https://github.com/fchollet/keras/issues/4339#issuecomment-260680102"}}}

**Comment by JustinYou1992** [comment]
@carlthome Yes. The real implementation depends  on the backend. Maybe I can calculate it approximately just like calculating the computational complexity of some network.

**Comment by JustinYou1992** [comment]
@carlthome Hi, I found a script here. [https://github.com/adri-romsor/FitNets/tree/master/scripts](url)

The 'computeNumberMult.py' and 'computeNumberParams.py' are to compute the number of multiplications and parameters in a model generated by 'pylearn2' instead of 'keras'.

**Comment by carlthome** [comment]
Ok? Good for you.

**Comment by jerinkantony** [comment]
Is there any solution for this?

**Comment by dav-ell** [comment]
Also interested.

**Comment by RichoMelbDs** [comment]
interested

**Comment by busyBeaver1** [comment]
I had the same problem. Here's my solution. It turns out that for a dense layer the number of multiplications should be the same as the number of weights. But this does not work as well for the convolution layers. For a convolution layer you need to multiply the number of weights by the number of instances of the kernel applied on the input to get the total number of multiplications. If my assumptions are right, then this should work for dense, conv and separable conv layers:

`import numpy as np

# here you need to create the model

print('\n'.join([
    'layer: ' + type(l).__name__ +\
    ', output shape: ' + str(l.output_shape[1:]) +\
    ', weights: ' + str(sum([i.size for i in l.get_weights()])) +\
    ', multiplications: ' + str(sum([i.size for i in l.get_weights()]) * int(np.prod(l.output_shape[1: -1])))\
    for l in model.layers
]))`

The output for my CNN (note that you should, at least temporarily, define the input shape):

`layer: SeparableConv2D, output shape: (240, 240, 128), weights: 400, multiplications: 23040000
layer: BatchNormalization, output shape: (240, 240, 128), weights: 512, multiplications: 29491200
layer: SeparableConv2D, output shape: (80, 80, 256), weights: 51456, multiplications: 329318400
layer: BatchNormalization, output shape: (80, 80, 256), weights: 1024, multiplications: 6553600
layer: SeparableConv2D, output shape: (40, 40, 256), weights: 90368, multiplications: 144588800
layer: BatchNormalization, output shape: (40, 40, 256), weights: 1024, multiplications: 1638400
layer: Dropout, output shape: (40, 40, 256), weights: 0, multiplications: 0
layer: SeparableConv2D, output shape: (40, 40, 256), weights: 84224, multiplications: 134758400
layer: Dropout, output shape: (40, 40, 256), weights: 0, multiplications: 0
layer: Conv2D, output shape: (40, 40, 5), weights: 92165, multiplications: 147464000`

It does not actually work for BatchNormalization layers, so the output for those layers probably is meaningless.
